// @sap/cds-dk/lib/init/template/enterprise-messaging-shared/index.js
const { join } = require('path')
const cds = require('../../../cds')
const { read } = cds.utils
const { readProject } = require('../../util/projectReader')
const { mergeJSON, mergeYAML } = require('../../util/merge')
const { srvNode4, srvJava4, enterpriseMessaging } = require('../_merging/registry-mta')

module.exports = class EnterpriseMessagingSharedTemplate extends require('../templateBase') {

  async canRun() {
    const { hasMta, hasHelm, isJava }  = await readProject(this.options)
    if (isJava)             throw `'cds add enterprise-messaging-shared' is not available for Java yet`
    if (hasHelm && !hasMta) throw `'cds add enterprise-messaging-shared' is not available for Kyma yet`
    return true
  }

  static hasFacet(env) {
    return env.requires?.messaging?.kind === 'enterprise-messaging-amqp'
  }

  async run() {
    const project = await readProject(this.options)
    const { configFile } = project
    project.shortcut = !(await read(configFile)).cds?.requires?.messaging?.kind
    await mergeJSON(configFile, join(__dirname, 'files', 'package.json.hbs'), project)
    await mergeJSON('event-mesh.json', join(__dirname, 'files', 'event-mesh.json.hbs'), project)
    await this.runDependentMerging()
  }

  async runDependentMerging() {
    const project = await readProject(this.options)
    const { hasMta, isNodejs, srvPath } = project

    if (hasMta) {
      const srv = (isNodejs ? srvNode4 : srvJava4)(srvPath)
      await mergeYAML(
          'mta.yaml',
          join(__dirname, 'files', 'mta.yaml.hbs'),
          project,
          {
              additions: [srv, enterpriseMessaging],
              relationships: [{
                  insert: [enterpriseMessaging, 'name'],
                  into: [srv, 'requires', 'name']
              }]
          }
      )
    }
  }
}
